# Project Outline: Data Combination
This is the data combination portion of the project "To Everything There is a Season : Using Weather Data and Demographic Information in the Predictive Modeling of Crimes in Dallas, Texas" by Ashley Steele.


[1. Importing Libraries and Setting Preferences](#1.-Importing-Libraries-and-Setting-Preferences)

[2. Importing Data Sources for Merging](#2.-Importing-Data-Sources-for-Merging)

[3. Combining Weather Data and Crime Data](#3.-Combining-Weather-Data-and-Crime-Data)

[4. Combining Census Data and Merged Weather/Crimes](#4.-Combining-Census-Data-and-Merged-Weather/Crimes)

[5. Saving Complete DF as a CSV](#5.-Saving-Complete-DF-as-a-CSV)

## 1. Importing Libraries and Setting Preferences
[Back to Outline](#Project-Outline:-Data-Combination)

In [6]:
# Importing all important libraries here!
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from datetime import datetime
from dateutil.parser import parse
import os

In [3]:
# Setting default number of decimal places for this notebook
pd.set_option('precision', 0)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [7]:
# Setting our plotting preferences
sns.set_style('whitegrid')
plt.figure(figsize= (20,20))
sns.set(palette= 'bright', font_scale = 1.5)

<Figure size 1440x1440 with 0 Axes>

In [8]:
# Checking directories 
os.getcwd()

'C:\\Users\\gothv\\Jupyter\\final_capstone'

# 2. Importing Data Sources for Merging
[Back to outline](#Project-Outline:-Data-Combination-and-Exploratory-Data-Analysis)

This project consists of three major individual data sets : Dallas Police Department crime calls from 2015 to 2018, daily weather information for the time period, and U.S. Census demographic data for the same time frame. We will be merging the three data sets into one final dataset below to use for further analysis and modeling.


In [9]:
# Importing dpd data
crimes= pd.read_csv('crimes_final.csv')

In [10]:
#Sanity check: Did it import correctly?
crimes.head()

Unnamed: 0               911_call_type  \
0       35350        11V - BURG MOTOR VEH   
1       35351      31 - CRIMINAL MISCHIEF   
2       35352  58 - ROUTINE INVESTIGATION   
3       35353                  40 - OTHER   
4       35354      31 - CRIMINAL MISCHIEF   

                                    type_of_incident   incident_address  \
0                                                BMV  4420 CEDARDALE DR   
1            CRIM MISCHIEF > OR EQUAL $50 BUT < $500  3071 SUNNYVALE ST   
2  FRAUD USE/POSS IDENTIFYING INFO-PRELIMINARY IN...    1400 S LAMAR ST   
3                             ASSAULT -VERBAL THREAT     9739 DENTON DR   
4            CRIM MISCHIEF > OR EQUAL $50 BUT < $500    10623 SYLVIA DR   

        division  sector council_district date_of_occurrence  \
0  SOUTH CENTRAL      73               D8         2015-01-01   
1  SOUTH CENTRAL      71               D4         2015-01-01   
2        CENTRAL      15               D2         2015-01-01   
3      NORTHWEST      52               D6         2015-01-01   
4      NORTHEAST      22               D9         2015-01-01   

   year_of_occurrence month_of_occurrence day1_of_the_week time_of_occurrence  \
0                2015             January              Thu              17:00   
1                2015             January              Thu              00:20   
2                2015             January              Thu              08:00   
3                2015             January              Thu              02:00   
4                2015             January              Thu              13:00   

   day_of_the_year victim_type         victim_race        victim_ethnicity  \
0                1  Individual               White  Non-Hispanic or Latino   
1                1  Individual               Black  Non-Hispanic or Latino   
2                1  Individual               Black  Non-Hispanic or Latino   
3                1  Individual  Hispanic or Latino      Hispanic or Latino   
4                1  Individual               Black  Non-Hispanic or Latino   

  victim_gender  victim_age responding_officer_#1__badge_no  \
0          Male          50                            8173   
1          Male          51                            8133   
2        Female          64                            7341   
3          Male          36                           10767   
4        Female          70                            9654   

  responding_officer_#1__name responding_officer_#2_badge_no  \
0      JONES,REGINALD,LADUNNE                            NaN   
1            ADAMS,CORY,JAMES                            NaN   
2             FREEMAN,DIANA,J                            NaN   
3                  HOVIS,ALAN                           5455   
4          BANDAS,WAYI,ALIBEY                            NaN   

  responding_officer_#2__name nibrs_crime_category  x_coordinate  y_cordinate  \
0                         NaN                  NaN         3e+06        7e+06   
1                         NaN                  NaN         2e+06        7e+06   
2                         NaN        MISCELLANEOUS         2e+06        7e+06   
3       AKON,FREDRICK,CHARLES                  NaN         2e+06        7e+06   
4                         NaN                  NaN         3e+06        7e+06   

   zip_code  year        date       mnth_yr  month  
0     75241  2015  2015-01-01  January-2015      1  
1     75216  2015  2015-01-01  January-2015      1  
2     75215  2015  2015-01-01  January-2015      1  
3     75220  2015  2015-01-01  January-2015      1  
4     75228  2015  2015-01-01  January-2015      1

In [11]:
# Checking crimes length
len(crimes)

255551

In [12]:
# Importing census data
census= pd.read_csv('census_final.csv')

In [13]:
# What does the census data look like?
census.head()

Unnamed: 0  total_pop  male  %_male  female  %_female  median_age  \
0           0       3951  2099      53    1852        47          40   
1           1       2907  1378      47    1529        53          37   
2           1       2907  1378      47    1529        53          37   
3           2       3759  1921      51    1838        49          34   
4           3       4316  2504      58    1812        42          34   

   18_&_over  21_&_over  62_&_over  65_&_over  %_white  %_black  %_native  \
0         80         78         12          9       90    6e+00        23   
1         78         77         16         10       96    4e-01         0   
2         78         77         16         10       96    4e-01         0   
3         82         82          8          6       89    3e+00         4   
4         85         85         12          8       93    1e+00        18   

   %_asian  %_hispanic  pop_over_16  %_pop_over_16  %_employed  %_unemployed  \
0    2e+00          10         3236           3236          72             3   
1    3e-01          17         2308           2308          72             1   
2    3e-01          17         2308           2308          72             1   
3    2e+00          18         3118           3118          83             2   
4    3e+00           6         3698           3698          79             0   

   mean_household_income  %_families_poverty  %_all_people_poverty  year  \
0                 183048                   3                     6  2015   
1                 133584                   0                     4  2015   
2                 133584                   0                     4  2015   
3                 112295                   2                     4  2015   
4                 150866                   0                     8  2015   

   zip_code  
0     75214  
1     75206  
2     75214  
3     75206  
4     75205

In [14]:
# Checking census length
len(census)

1347

In [15]:
# Importing weather data
weather = pd.read_csv('weather_final.csv')

In [16]:
# What does this data look like?
weather.head()

date  temp_in_F  humidity  percip_inches  year  month  day
0  2015-01-01         34        89          3e-02  2015      1    1
1  2015-01-02         39        97          2e-02  2015      1    2
2  2015-01-03         44        85          2e-02  2015      1    3
3  2015-01-04         35        67          0e+00  2015      1    4
4  2015-01-05         36        61          0e+00  2015      1    5

In [17]:
# Checking our weather length
len(weather)

1448

Combining, or joining, these data sets will be interesting as they do not all share a common key. Let's take this step by step, first starting with combining our weather and crime data!

## 3. Combining Weather Data and Crime Data
[Back to outline](#Project-Outline:-Data-Combination-and-Exploratory-Data-Analysis)

Before we can join these two datasets we need to do a little bit of house-keeping to ensure that our data goes together in the best possible way.

In [18]:
# Dropping year, month, and day columns in our weather df
weather.drop(axis= 1, columns = ['year', 'month', 'day'], inplace = True)

In [19]:
# Converting weather['date'] back  into a datetime 
weather['date'] = pd.to_datetime(weather['date'])

In [20]:
# Creating a date only feature for later use
weather['date_only'] = weather['date'].dt.date 

In [21]:
# Sanity check: Does it look like we expect?
weather.head()

date  temp_in_F  humidity  percip_inches   date_only
0 2015-01-01         34        89          3e-02  2015-01-01
1 2015-01-02         39        97          2e-02  2015-01-02
2 2015-01-03         44        85          2e-02  2015-01-03
3 2015-01-04         35        67          0e+00  2015-01-04
4 2015-01-05         36        61          0e+00  2015-01-05

In [22]:
# Converting date in crimes to a datetime object to match our weather date
crimes['date']= pd.to_datetime(crimes['date'])

In [23]:
# Making a new key to match our datasets on for merging!
crimes['date_only'] = crimes['date'].dt.date

In [24]:
# Getting rid of the few rows in crimes that don't have a zip code associated with them
crimes.dropna(axis = 0, subset = ['zip_code'], inplace = True)

In [25]:
# Joining weather and crimes on date
final_df = pd.merge(crimes, weather, how = 'left', left_on= 'date_only', right_on ='date_only')

In [26]:
# Checking length of final so far 
len(final_df)

255551

In [27]:
# Comparing to original crimes length
len(crimes)

255551

Excellent! Our before and after lengths are the same! Now let's take a closer look at any weird outliers or missing data we may have before we go on!

In [28]:
# What do our nulls look like, if any?
final_df.isnull().sum().sort_values(ascending = False)

responding_officer_#2__name        170133
responding_officer_#2_badge_no     170133
nibrs_crime_category               127873
percip_inches                        2407
temp_in_F                            2407
date_y                               2407
humidity                             2407
council_district                      176
division                               41
sector                                 41
time_of_occurrence                      0
day1_of_the_week                        0
month_of_occurrence                     0
date_of_occurrence                      0
year_of_occurrence                      0
victim_type                             0
incident_address                        0
type_of_incident                        0
911_call_type                           0
day_of_the_year                         0
victim_age                              0
victim_race                             0
victim_ethnicity                        0
victim_gender                     

The first time we ran the cell above we had 2,407 records without weather information. Why? It seems that on several days we had weather information with odd errors. Let's go ahead and fix these now!

In [29]:
# Let's use a moving average to help us fill in these odd/missing values with the average of values around them 
ema_df = final_df.iloc[:,-3:].ewm(span=14,adjust=False).mean()
ema_df.head()

temp_in_F  humidity  percip_inches
0         34        89          3e-02
1         34        89          3e-02
2         34        89          3e-02
3         34        89          3e-02
4         34        89          3e-02

In [30]:
# Replacing the null weather values with the moving average values
final_df['temp_in_F']= final_df['temp_in_F'].fillna(ema_df['temp_in_F'])
final_df['humidity']= final_df['humidity'].fillna(ema_df['humidity'])
final_df['percip_inches']= final_df['percip_inches'].fillna(ema_df['percip_inches'])

In [31]:
# Double checking again for nulls
final_df.isnull().sum().sort_values(ascending = False)

responding_officer_#2_badge_no     170133
responding_officer_#2__name        170133
nibrs_crime_category               127873
date_y                               2407
council_district                      176
division                               41
sector                                 41
percip_inches                           0
day_of_the_year                         0
time_of_occurrence                      0
day1_of_the_week                        0
date_of_occurrence                      0
month_of_occurrence                     0
year_of_occurrence                      0
victim_race                             0
incident_address                        0
type_of_incident                        0
911_call_type                           0
victim_type                             0
victim_age                              0
victim_ethnicity                        0
victim_gender                           0
humidity                                0
responding_officer_#1__badge_no   

We now have values for our missing data! Taking one quick look at how our final_df looks before going on.

In [32]:
# Sanity Check: Does it all look ok?
final_df.head()

Unnamed: 0               911_call_type  \
0       35350        11V - BURG MOTOR VEH   
1       35351      31 - CRIMINAL MISCHIEF   
2       35352  58 - ROUTINE INVESTIGATION   
3       35353                  40 - OTHER   
4       35354      31 - CRIMINAL MISCHIEF   

                                    type_of_incident   incident_address  \
0                                                BMV  4420 CEDARDALE DR   
1            CRIM MISCHIEF > OR EQUAL $50 BUT < $500  3071 SUNNYVALE ST   
2  FRAUD USE/POSS IDENTIFYING INFO-PRELIMINARY IN...    1400 S LAMAR ST   
3                             ASSAULT -VERBAL THREAT     9739 DENTON DR   
4            CRIM MISCHIEF > OR EQUAL $50 BUT < $500    10623 SYLVIA DR   

        division  sector council_district date_of_occurrence  \
0  SOUTH CENTRAL      73               D8         2015-01-01   
1  SOUTH CENTRAL      71               D4         2015-01-01   
2        CENTRAL      15               D2         2015-01-01   
3      NORTHWEST      52               D6         2015-01-01   
4      NORTHEAST      22               D9         2015-01-01   

   year_of_occurrence month_of_occurrence day1_of_the_week time_of_occurrence  \
0                2015             January              Thu              17:00   
1                2015             January              Thu              00:20   
2                2015             January              Thu              08:00   
3                2015             January              Thu              02:00   
4                2015             January              Thu              13:00   

   day_of_the_year victim_type         victim_race        victim_ethnicity  \
0                1  Individual               White  Non-Hispanic or Latino   
1                1  Individual               Black  Non-Hispanic or Latino   
2                1  Individual               Black  Non-Hispanic or Latino   
3                1  Individual  Hispanic or Latino      Hispanic or Latino   
4                1  Individual               Black  Non-Hispanic or Latino   

  victim_gender  victim_age responding_officer_#1__badge_no  \
0          Male          50                            8173   
1          Male          51                            8133   
2        Female          64                            7341   
3          Male          36                           10767   
4        Female          70                            9654   

  responding_officer_#1__name responding_officer_#2_badge_no  \
0      JONES,REGINALD,LADUNNE                            NaN   
1            ADAMS,CORY,JAMES                            NaN   
2             FREEMAN,DIANA,J                            NaN   
3                  HOVIS,ALAN                           5455   
4          BANDAS,WAYI,ALIBEY                            NaN   

  responding_officer_#2__name nibrs_crime_category  x_coordinate  y_cordinate  \
0                         NaN                  NaN         3e+06        7e+06   
1                         NaN                  NaN         2e+06        7e+06   
2                         NaN        MISCELLANEOUS         2e+06        7e+06   
3       AKON,FREDRICK,CHARLES                  NaN         2e+06        7e+06   
4                         NaN                  NaN         3e+06        7e+06   

   zip_code  year     date_x       mnth_yr  month   date_only     date_y  \
0     75241  2015 2015-01-01  January-2015      1  2015-01-01 2015-01-01   
1     75216  2015 2015-01-01  January-2015      1  2015-01-01 2015-01-01   
2     75215  2015 2015-01-01  January-2015      1  2015-01-01 2015-01-01   
3     75220  2015 2015-01-01  January-2015      1  2015-01-01 2015-01-01   
4     75228  2015 2015-01-01  January-2015      1  2015-01-01 2015-01-01   

   temp_in_F  humidity  percip_inches  
0         34        89          3e-02  
1         34        89          3e-02  
2         34        89          3e-02  
3         34        89          3e-02  
4         34        89          

In [33]:
# While looking at this dataset in closer detail I realized that the "year_of_occurrence" column potentially has errors, 
# so let's drop it
final_df.drop(columns = 'year_of_occurrence', inplace = True)

In [34]:
# Exporting this dataset to csv
final_df.to_csv('final_df.csv')

Excellent! The only remaining null values in this dataset are ones we expected to see!

Now that we have our weather and crimes combined we can move on to the next step: adding our census data to our final data frame!

## 4. Combining Census Data and Merged Weather/Crimes
[Back to Outline](#Project-Outline:-Data-Combination)

Since the key we have in common in these two datasets is zip code we need to look at it (and anything relevant) closely before attempting to merge.

In [35]:
# Since we've merged two seperate datasets let's make sure our data is still looking like we expect
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 255551 entries, 0 to 255550
Data columns (total 34 columns):
Unnamed: 0                         255551 non-null int64
911_call_type                      255551 non-null object
type_of_incident                   255551 non-null object
incident_address                   255551 non-null object
division                           255510 non-null object
sector                             255510 non-null float64
council_district                   255375 non-null object
date_of_occurrence                 255551 non-null object
month_of_occurrence                255551 non-null object
day1_of_the_week                   255551 non-null object
time_of_occurrence                 255551 non-null object
day_of_the_year                    255551 non-null int64
victim_type                        255551 non-null object
victim_race                        255551 non-null object
victim_ethnicity                   255551 non-null object
victim_gender       

In [36]:
# Do we have any weird columns after our merge?
final_df.columns

Index(['Unnamed: 0', '911_call_type', 'type_of_incident', 'incident_address',
       'division', 'sector', 'council_district', 'date_of_occurrence',
       'month_of_occurrence', 'day1_of_the_week', 'time_of_occurrence',
       'day_of_the_year', 'victim_type', 'victim_race', 'victim_ethnicity',
       'victim_gender', 'victim_age', 'responding_officer_#1__badge_no',
       'responding_officer_#1__name', 'responding_officer_#2_badge_no',
       'responding_officer_#2__name', 'nibrs_crime_category', 'x_coordinate',
       'y_cordinate', 'zip_code', 'year', 'date_x', 'mnth_yr', 'month',
       'date_only', 'date_y', 'temp_in_F', 'humidity', 'percip_inches'],
      dtype='object')

In [37]:
#Dropping our date columns
final_df.drop(columns = ['date_x', 'date_y'], inplace = True)

In [38]:
# What does our census info look like overall?
census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1347 entries, 0 to 1346
Data columns (total 25 columns):
Unnamed: 0               1347 non-null int64
total_pop                1347 non-null int64
male                     1347 non-null int64
%_male                   1347 non-null float64
female                   1347 non-null int64
%_female                 1347 non-null float64
median_age               1347 non-null float64
18_&_over                1347 non-null float64
21_&_over                1347 non-null float64
62_&_over                1347 non-null float64
65_&_over                1347 non-null float64
%_white                  1347 non-null float64
%_black                  1347 non-null float64
%_native                 1347 non-null int64
%_asian                  1347 non-null float64
%_hispanic               1347 non-null float64
pop_over_16              1347 non-null int64
%_pop_over_16            1347 non-null int64
%_employed               1347 non-null float64
%_unemployed  

In [39]:
# What dtype is zipcode in our final df?
final_df['zip_code'].describe()

count    255551
mean      75225
std         172
min           0
25%       75215
50%       75224
75%       75237
max       75390
Name: zip_code, dtype: float64

In [40]:
# What dtype is our zip code in our census data?
census['zip_code'].describe()

count     1347
mean     75225
std         17
min      75201
25%      75214
50%      75225
75%      75235
max      75390
Name: zip_code, dtype: float64

In [41]:
# Converting our floats to intergers to get whole numbers
final_df['zip_code'] = final_df['zip_code'].astype(int)
census['zip_code'] = census['zip_code'].astype(int)

Awesome! Both of our zip codes are now intergers which means, hopefully, that we should have little to no problem combining datasets on this key!Let's try it now!

In [42]:
# Since we have columns with the same name in our new and old df (that aren't our key) we have to get rid of the duplicates
final_df.drop(columns = 'month_of_occurrence', inplace = True)

In [43]:
# Making a copy of the final to merge on, just incase
final1_df= final_df.copy()

In [44]:
# Checking length of final1
len(final1_df)

255551

In [45]:
# Double checking to make sure we only have data up to 2018
final1_df.loc[final1_df['year'] == 2019]

Empty DataFrame
Columns: [Unnamed: 0, 911_call_type, type_of_incident, incident_address, division, sector, council_district, date_of_occurrence, day1_of_the_week, time_of_occurrence, day_of_the_year, victim_type, victim_race, victim_ethnicity, victim_gender, victim_age, responding_officer_#1__badge_no, responding_officer_#1__name, responding_officer_#2_badge_no, responding_officer_#2__name, nibrs_crime_category, x_coordinate, y_cordinate, zip_code, year, mnth_yr, month, date_only, temp_in_F, humidity, percip_inches]
Index: []

In [46]:
# Converting year and zip to a string
final1_df['zip_code'] = final1_df['zip_code'].astype(str)
final1_df['year'] = final1_df['year'].astype(str)

In [47]:
# Let's see if we can make a zipcode and year combined column to make our merging easier
final1_df['zip_year']= final1_df['zip_code'] + '- ' + final1_df['year']

In [48]:
# Checking what our df looks like so far
final1_df.head()

Unnamed: 0               911_call_type  \
0       35350        11V - BURG MOTOR VEH   
1       35351      31 - CRIMINAL MISCHIEF   
2       35352  58 - ROUTINE INVESTIGATION   
3       35353                  40 - OTHER   
4       35354      31 - CRIMINAL MISCHIEF   

                                    type_of_incident   incident_address  \
0                                                BMV  4420 CEDARDALE DR   
1            CRIM MISCHIEF > OR EQUAL $50 BUT < $500  3071 SUNNYVALE ST   
2  FRAUD USE/POSS IDENTIFYING INFO-PRELIMINARY IN...    1400 S LAMAR ST   
3                             ASSAULT -VERBAL THREAT     9739 DENTON DR   
4            CRIM MISCHIEF > OR EQUAL $50 BUT < $500    10623 SYLVIA DR   

        division  sector council_district date_of_occurrence day1_of_the_week  \
0  SOUTH CENTRAL      73               D8         2015-01-01              Thu   
1  SOUTH CENTRAL      71               D4         2015-01-01              Thu   
2        CENTRAL      15               D2         2015-01-01              Thu   
3      NORTHWEST      52               D6         2015-01-01              Thu   
4      NORTHEAST      22               D9         2015-01-01              Thu   

  time_of_occurrence  day_of_the_year victim_type         victim_race  \
0              17:00                1  Individual               White   
1              00:20                1  Individual               Black   
2              08:00                1  Individual               Black   
3              02:00                1  Individual  Hispanic or Latino   
4              13:00                1  Individual               Black   

         victim_ethnicity victim_gender  victim_age  \
0  Non-Hispanic or Latino          Male          50   
1  Non-Hispanic or Latino          Male          51   
2  Non-Hispanic or Latino        Female          64   
3      Hispanic or Latino          Male          36   
4  Non-Hispanic or Latino        Female          70   

  responding_officer_#1__badge_no responding_officer_#1__name  \
0                            8173      JONES,REGINALD,LADUNNE   
1                            8133            ADAMS,CORY,JAMES   
2                            7341             FREEMAN,DIANA,J   
3                           10767                  HOVIS,ALAN   
4                            9654          BANDAS,WAYI,ALIBEY   

  responding_officer_#2_badge_no responding_officer_#2__name  \
0                            NaN                         NaN   
1                            NaN                         NaN   
2                            NaN                         NaN   
3                           5455       AKON,FREDRICK,CHARLES   
4                            NaN                         NaN   

  nibrs_crime_category  x_coordinate  y_cordinate zip_code  year  \
0                  NaN         3e+06        7e+06    75241  2015   
1                  NaN         2e+06        7e+06    75216  2015   
2        MISCELLANEOUS         2e+06        7e+06    75215  2015   
3                  NaN         2e+06        7e+06    75220  2015   
4                  NaN         3e+06        7e+06    75228  2015   

        mnth_yr  month   date_only  temp_in_F  humidity  percip_inches  \
0  January-2015      1  2015-01-01         34        89          3e-02   
1  January-2015      1  2015-01-01         34        89          3e-02   
2  January-2015      1  2015-01-01         34        89          3e-02   
3  January-2015      1  2015-01-01         34        89          3e-02   
4  January-2015      1  2015-01-01         34        89          3e-02   

      zip_year  
0  75241- 2015  
1  75216- 2015  
2  75215- 2015  
3  75220- 2015  
4  75228- 2015

In [49]:
# Making year and zipcode strings in census df
census['year'] = census['year'].astype(str)
census['zip_code'] = census['zip_code'].astype(str)

In [50]:
# Making a key to merge on in our census data
census['zip_year']= census['zip_code'] + '-' + census['year']

We now know that any resulting data frame we create needs to have the same number of rows, since our first merge was the least intensive!

In [51]:
# Adding census data to our final df
complete = pd.merge(final1_df, census, on = ['zip_code', 'year'], how = 'left' )

In [52]:
# How long is this new df?
len(complete)

2313408

In [53]:
# Checking for nulls in our new df
complete.isnull().sum().sort_values(ascending = False)

responding_officer_#2_badge_no     1525684
responding_officer_#2__name        1525684
nibrs_crime_category               1514046
zip_year_y                           65731
62_&_over                            65731
%_all_people_poverty                 65731
Unnamed: 0_y                         65731
total_pop                            65731
male                                 65731
female                               65731
%_female                             65731
median_age                           65731
18_&_over                            65731
21_&_over                            65731
%_male                               65731
65_&_over                            65731
pop_over_16                          65731
%_families_poverty                   65731
mean_household_income                65731
%_unemployed                         65731
%_white                              65731
%_pop_over_16                        65731
%_employed                           65731
%_hispanic 

In [54]:
# Getting rid of our null values after our merge
complete.dropna(subset = ['total_pop'], inplace = True)

We now have census data for all of our crimes, but it appears we have duplicate records. Let's see if we can get rid of these.

In [55]:
# How many unique values do we have in our final1 dataset?
final1_df['Unnamed: 0'].nunique()

255551

In [56]:
# How much did these change after merging?
complete['Unnamed: 0_x'].nunique()

189820

In [58]:
# Using the index of our first dataframe to get ride of any duplicates
complete.drop_duplicates(subset= 'Unnamed: 0_x', keep='first', inplace=True)

In [59]:
# What is our new lenght?
len(complete)

189820

In [60]:
# Double checking to see if nulls have been dealt with!
complete.isnull().sum().sort_values(ascending= False)

nibrs_crime_category               127576
responding_officer_#2__name        126647
responding_officer_#2_badge_no     126647
council_district                      141
division                                9
sector                                  9
zip_year_y                              0
victim_age                              0
mnth_yr                                 0
year                                    0
zip_code                                0
y_cordinate                             0
x_coordinate                            0
responding_officer_#1__name             0
responding_officer_#1__badge_no         0
victim_ethnicity                        0
victim_gender                           0
%_all_people_poverty                    0
victim_race                             0
victim_type                             0
day_of_the_year                         0
time_of_occurrence                      0
day1_of_the_week                        0
date_of_occurrence                

HOOOOOOOOOORRRRAAAAAYYYYY!! We finally have our ugly null values gone! Let's make this new dataframe prettier!

In [61]:
# Getting rid of the unnammed column & multiple columns
complete.drop(columns = ['Unnamed: 0_x', 'zip_year_x', 'Unnamed: 0_y'], inplace = True)

In [62]:
# Renaming what was left of the multiples
complete.rename(columns = {'zip_year_y':'zip_year'}, inplace = True)

In [63]:
# Checking that our renames worked
complete.head()

911_call_type  \
0         11V - BURG MOTOR VEH   
10      31 - CRIMINAL MISCHIEF   
28  58 - ROUTINE INVESTIGATION   
37                  40 - OTHER   
49      31 - CRIMINAL MISCHIEF   

                                     type_of_incident   incident_address  \
0                                                 BMV  4420 CEDARDALE DR   
10            CRIM MISCHIEF > OR EQUAL $50 BUT < $500  3071 SUNNYVALE ST   
28  FRAUD USE/POSS IDENTIFYING INFO-PRELIMINARY IN...    1400 S LAMAR ST   
37                             ASSAULT -VERBAL THREAT     9739 DENTON DR   
49            CRIM MISCHIEF > OR EQUAL $50 BUT < $500    10623 SYLVIA DR   

         division  sector council_district date_of_occurrence  \
0   SOUTH CENTRAL      73               D8         2015-01-01   
10  SOUTH CENTRAL      71               D4         2015-01-01   
28        CENTRAL      15               D2         2015-01-01   
37      NORTHWEST      52               D6         2015-01-01   
49      NORTHEAST      22               D9         2015-01-01   

   day1_of_the_week time_of_occurrence  day_of_the_year victim_type  \
0               Thu              17:00                1  Individual   
10              Thu              00:20                1  Individual   
28              Thu              08:00                1  Individual   
37              Thu              02:00                1  Individual   
49              Thu              13:00                1  Individual   

           victim_race        victim_ethnicity victim_gender  victim_age  \
0                White  Non-Hispanic or Latino          Male          50   
10               Black  Non-Hispanic or Latino          Male          51   
28               Black  Non-Hispanic or Latino        Female          64   
37  Hispanic or Latino      Hispanic or Latino          Male          36   
49               Black  Non-Hispanic or Latino        Female          70   

   responding_officer_#1__badge_no responding_officer_#1__name  \
0                             8173      JONES,REGINALD,LADUNNE   
10                            8133            ADAMS,CORY,JAMES   
28                            7341             FREEMAN,DIANA,J   
37                           10767                  HOVIS,ALAN   
49                            9654          BANDAS,WAYI,ALIBEY   

   responding_officer_#2_badge_no responding_officer_#2__name  \
0                             NaN                         NaN   
10                            NaN                         NaN   
28                            NaN                         NaN   
37                           5455       AKON,FREDRICK,CHARLES   
49                            NaN                         NaN   

   nibrs_crime_category  x_coordinate  y_cordinate zip_code  year  \
0                   NaN         3e+06        7e+06    75241  2015   
10                  NaN         2e+06        7e+06    75216  2015   
28        MISCELLANEOUS         2e+06        7e+06    75215  2015   
37                  NaN         2e+06        7e+06    75220  2015   
49                  NaN         3e+06        7e+06    75228  2015   

         mnth_yr  month   date_only  temp_in_F  humidity  percip_inches  \
0   January-2015      1  2015-01-01         34        89          3e-02   
10  January-2015      1  2015-01-01         34        89          3e-02   
28  January-2015      1  2015-01-01         34        89          3e-02   
37  January-2015      1  2015-01-01         34        89          3e-02   
49  January-2015      1  2015-01-01         34        89          3e-02   

    total_pop  male  %_male  female  %_female  median_age  18_&_over  \
0         942   357      38     585        62          27         63   
10       4410  2126      48    2284        52          33         66   
28       1413   772      55     641        45          37         79   
37       7966  4243      53    3723        47          31         76   
49       3527  1792      51    1735        49          30         

Our data looks good so far, but let's go ahead and re-organize our columns to make a bit more sense flow-wise before we do some exploring!

In [64]:
# Changing the order of our columns to make our data a little more readable!
complete = complete[['date_only','month', 'mnth_yr', 'day1_of_the_week','time_of_occurrence', 'day_of_the_year', 'temp_in_F',
'humidity', 'percip_inches', '911_call_type', 'type_of_incident', 'division', 'sector','council_district', 'victim_type',
'victim_race', 'victim_ethnicity', 'victim_gender', 'victim_age', 'responding_officer_#1__badge_no', 'responding_officer_#1__name',
'responding_officer_#2_badge_no', 'responding_officer_#2__name', 'x_coordinate', 'y_cordinate', 'zip_code', 'total_pop', 
'male', '%_male', 'female', '%_female', 'median_age', '18_&_over', '21_&_over', '62_&_over', '65_&_over', '%_white', 
'%_black', '%_native', '%_asian','%_hispanic', 'pop_over_16', '%_pop_over_16', '%_employed', '%_unemployed', 
'mean_household_income', '%_families_poverty', '%_all_people_poverty', 'year', 'zip_code']]

Now that our data is FINALLY in a format that we can use let's save the information to a csv file so we can do some exploration in our next notebook!

## 5. Saving Complete DF as a CSV
[Back to Outline](#Project-Outline:-Data-Combination)

In [65]:
# Now that we have done all of our merging and cleaning of this data let's save a copy of this dataframe as a csv for later use!
complete.to_csv('complete.csv')

Want to know what I did with this data? Check out the project page [here](https://steelealloy.github.io/final_capstone/)